pose_broadcaster/src/pose_broadcaster.cpp

```diff
@@ -0,0 +1,195 @@
+
+  try
+  {
+    pose_publisher_ = get_node()->create_publisher<geometry_msgs::msg::PoseStamped>(
+      DEFAULT_POSE_TOPIC, rclcpp::SystemDefaultsQoS());
+    realtime_publisher_ =
+      std::make_unique<realtime_tools::RealtimePublisher<geometry_msgs::msg::PoseStamped>>(
+        pose_publisher_);
+
+    if (params_.tf.enable)
+    {
+      tf_publisher_ = get_node()->create_publisher<tf2_msgs::msg::TFMessage>(
+        DEFAULT_TF_TOPIC, rclcpp::SystemDefaultsQoS());
+      realtime_tf_publisher_ =
+        std::make_unique<realtime_tools::RealtimePublisher<tf2_msgs::msg::TFMessage>>(
+          tf_publisher_);
+    }
+  }
+  catch (const std::exception & ex)
+  {
+    fprintf(
+      stderr, "Exception thrown during publisher creation at configure stage with message: %s\n",
+      ex.what());
+    return controller_interface::CallbackReturn::ERROR;
+  }
+
+  // Initialize pose message
+  realtime_publisher_->lock();
+  realtime_publisher_->msg_.header.frame_id = params_.frame_id;
+  realtime_publisher_->unlock();
+
+  // Initialize tf message if tf publishing is enabled
+  if (realtime_tf_publisher_)
+  {
+    realtime_tf_publisher_->lock();
+
+    realtime_tf_publisher_->msg_.transforms.resize(1);
+    auto & tf_transform = realtime_tf_publisher_->msg_.transforms.front();
+    tf_transform.header.frame_id = params_.frame_id;
+    if (params_.tf.child_frame_id.empty())
+    {
+      tf_transform.child_frame_id = params_.pose_name;
+    }
+    else
+    {
+      tf_transform.child_frame_id = params_.tf.child_frame_id;
+    }
+
+    realtime_tf_publisher_->unlock();
+  }
+
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+controller_interface::CallbackReturn PoseBroadcaster::on_activate(
+  const rclcpp_lifecycle::State & /*previous_state*/)
+{
+  pose_sensor_->assign_loaned_state_interfaces(state_interfaces_);
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+controller_interface::CallbackReturn PoseBroadcaster::on_deactivate(
+  const rclcpp_lifecycle::State & /*previous_state*/)
+{
+  pose_sensor_->release_interfaces();
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+controller_interface::return_type PoseBroadcaster::update(
+  const rclcpp::Time & time, const rclcpp::Duration & /*period*/)
+{
+  geometry_msgs::msg::Pose pose;
+  pose_sensor_->get_values_as_message(pose);
+
+  if (realtime_publisher_ && realtime_publisher_->trylock())
+  {
+    realtime_publisher_->msg_.header.stamp = time;
+    realtime_publisher_->msg_.pose = pose;
+    realtime_publisher_->unlockAndPublish();
+  }
+
+  if (realtime_tf_publisher_ && realtime_tf_publisher_->trylock())
+  {
+    bool do_publish = false;
+    // rlcpp::Time comparisons throw if clock types are not the same
+    if (tf_last_publish_time_.get_clock_type() != time.get_clock_type())
+    {
+      do_publish = true;
+    }
+    else if (!tf_publish_period_ || (tf_last_publish_time_ + *tf_publish_period_ < time))
```

Comment:
If we use ` < time`, the tf would only be published every second update if the controller rate is exactly the same as the tf publish rate, right?
```suggestion
    else if (!tf_publish_period_ || (tf_last_publish_time_ + *tf_publish_period_ <= time))
```

Reply:
Good point, i didn't think of that